### PR TITLE
Make it possible to join pseudo ID rooms

### PR DESCRIPTION
**Pull request by S7evinK**
This should allow us to join pseudo ID rooms over federation.
Also needs https://github.com/matrix-org/gomatrixserverlib/pull/397

### CODE REVIEW

File: docs/installation/manual/3_signingkey.md
Comment (devonh) [inline docs/installation/manual/3_signingkey.md:1]: why is this change pulled into this PR?

File: docs/installation/manual/4_configuration.md
Comment (devonh) [inline docs/installation/manual/4_configuration.md:1]: why is this change pulled into this PR?

File: roomserver/internal/perform/perform_join.go [hunk 1]
```diff
@@ -314,7 +314,7 @@ func (r *Joiner) performJoinRoomByID(
 		// sign the event with the pseudo ID key
 		identity = fclient.SigningIdentity{
 			ServerName: "self",
-			KeyID:      "ed25519",
+			KeyID:      "ed25519:1",
```
Comment (devonh) [inline roomserver/internal/perform/perform_join.go:317]: what does the 1 mean?

File: roomserver/storage/sqlite3/user_room_keys_table.go [hunk 2]
```diff
@@ -137,7 +138,7 @@ func (s *userRoomKeysStatements) BulkSelectUserNIDs(ctx context.Context, txn *sq
 	selectSQL := strings.Replace(selectUserNIDsSQL, "($2)", sqlutil.QueryVariadicOffset(len(senders), len(senderKeys)), 1)
 	selectSQL = strings.Replace(selectSQL, "($1)", sqlutil.QueryVariadic(len(senderKeys)), 1) // replace $1 with the roomNIDs
 
-	selectStmt, err := txn.Prepare(selectSQL)
+	selectStmt, err := s.db.Prepare(selectSQL)
```
Comment (devonh) [inline roomserver/storage/sqlite3/user_room_keys_table.go:141]: why this change / did this not work before?

File: roomserver/storage/shared/storage.go [hunk 3]
```diff
@@ -1757,58 +1762,54 @@ func (d *Database) SelectUserRoomPublicKey(ctx context.Context, userID spec.User
 // SelectUserIDsForPublicKeys returns a map from roomID -> map from senderKey -> userID
 func (d *Database) SelectUserIDsForPublicKeys(ctx context.Context, publicKeys map[spec.RoomID][]ed25519.PublicKey) (result map[spec.RoomID]map[string]string, err error) {
 	result = make(map[spec.RoomID]map[string]string, len(publicKeys))
-	err = d.Writer.Do(d.DB, nil, func(txn *sql.Tx) error {
```
Comment (devonh) [inline roomserver/storage/shared/storage.go:1760]: curious why move this out of the `Do` block

File: docs/installation/manual/3_signingkey.md
Comment (S7evinK) [reply docs/installation/manual/3_signingkey.md:1]: This branch has pulled main, while the other did not.

File: roomserver/internal/perform/perform_join.go (hunk 1 above)
Comment (S7evinK) [reply roomserver/internal/perform/perform_join.go:317]: We have a `isValidID` in GMSL, which checks for a `:`, so added this. It's basically unused.

File: roomserver/storage/shared/storage.go (hunk 3 above)
Comment (S7evinK) [reply roomserver/storage/shared/storage.go:1760]: With SQLite this would deadlock, as there may be another writer waiting to create a key in the DB. And since we don't write anything in this method, it should be safe without a writer.

File: roomserver/storage/sqlite3/user_room_keys_table.go (hunk 2 above)
Comment (S7evinK) [reply roomserver/storage/sqlite3/user_room_keys_table.go:141]: It did work, while we had a writer, which we now don't have, given the change below. Maybe better to use `sqlutil.TxStmt`

**Review by devonh — APPROVED** [review]
